Start a Tutorial how to implement a Solver. (#255)
JuliaManifolds/Manopt.jl: tutorials/ImplementASolver.qmd

    3. If $f(p^{(k+1)}) < f(q)$ set q = p^{(k+1)}$ as our new best visited point
4. Return $q$ as the resulting best point we visited

## Preliminaries – Elements a Solver works on

THere is two main ingredients a solver needs:
A problem to work on and the state of a solver, which “identifies” the solver.

### The “Task” – An `AbstractManoptProblem`

A problem in `Manopt.jl` usually consists of a manifold (an [`AbstractManifold`](https://juliamanifolds.github.io/ManifoldsBase.jl/stable/types.html#The-AbstractManifold)) and an [`AbstractManifoldObjective`](@ref)
describing the function we have and its features.
In our case the objective is (just) a [`ManifoldCostObjective`](@ref) that stores cost function
`f(M,p) = ...`. MOre general it might for example store a gradient function or the Hessian
or any other information we have about our task.

This is something independent of the solver itself, since it only identifies the problem we
want to solve independent of how we want to solve it – or in other words, this type contains
all information that is static and independent of the specific solver at hand.

Usually the problems variable is called `mp`.

> ### The “Solver” An `AbstractManoptSolverState`

Author: “ – an” is probably the best choice so

The Solver – An AbstractManoptSolverState

the retraction based on a given manifold.

```{julia}
#| output: false
function RandomWalkState(M::AbstractManifold, p::P=rand(M);
    σ = 0.1,
    retraction_method::R=default_retraction_method(M),
    stopping_criterion::S=StopAfterIteration(200)
) where {P, R<:AbstractRetractionMethod, S<:StoppingCriterion}
    return RandomWalkState{P,R,S}(p, copy(M, p), σ, retraction_method, stopping_criterion)
end
```

Parametrising the state avoid that we have abstract typed fields.
The keyword arguments for the retraction and stopping criterion are the ones usually used
in `Manopt.jl` and provide an easy way to construct this state now.

States usually have a shortened name as their variable, we will use `rws` for our state here.

## Implementing the Your solver

There is basically only two methods we need to implement for our solver

- `initialize_solver!(mp, rws)` which initialises the solver before the first iteration
- `step_solver!(mp, rws, i)` implement the `i`th iteration, where `i` is given to you as the third parameter
- `get_iterate(rws)` to access the iterate from other places in the solver
- `get_solver_result(rws)` toi return the final point we reached. Bz default this would

>   return the last iterate, but since our result is `q`, we have to set this

Author: To be honest I wrote that and the code at the same time, so I even have to read it carefully myself ;)

```{julia}
#| output: false
function initialize_solver!(mp::AbstractManoptProblem, rws::RandomWalkState)
    copyto!(M, rws.q, rws.p) # Set p^{(0)} = q
    return rws
end
get_iterate(rws::RandomWalkState) = rws.p
get_solver_result(rws::RandomWalkState) = rws.q
```

and siilarly we implement the step. Here we make use of the fact that the problem
(and also the objective in fact) have access functions for their elements,
the one we need is [`get_cost`](@ref).

```{julia}
#| output: false
function step_solver!(mp::AbstractManoptProblem, rws::RandomWalkState, i)
    M = get_manifold(mp) # for ease of use get the manifold from the problem
    X = rand(M; vector_at=p)     # generate a direction
    X .*= rws.σ/norm(M, p, X)
    # Walk
    retract!(M, rws.p, rws.p, X, rws.retraction_method)
    # is the new point better? Then store it
    if get_cost(mp, rws.p) < get_cost(mp, rws.q)
        copyto!(M, rws.p, rws.q)
    end
    return rws
end
```

Performance wise we could improve the number of allocations by making `X` also a field of
our `rws` but let's keep it simple here.
We could also store the cost of `q` in the state, but we will see how to easily also enable

> this solver to allow for [caching](https://manoptjl.org/stable/tutorials/CountAndCache/#How-to-Count-and-Cache-Function-Calls).

Reviewer: ```suggestion
this solver to allow for [caching](https://manoptjl.org/stable/tutorials/CountAndCache/#How-to-Count-and-Cache-Function-Calls).
In practice, however, it is preferable to cache intermediate values like cost of `q` in the state when it can be easily achieved. This way we do not have to deal with overheads of an external cache.
```

Now we can just run the solver already! We take the same example as for the other tutorials

We first define our task, the Riemannian Center of Mass from the [Get Started: Optimize!](https://manoptjl.org/stable/tutorials/Optimize!.html) tutorial.

```{julia}
#| output: false
Random.seed!(23)
n = 100
σ = π / 8
M = Sphere(2)
p = 1 / sqrt(2) * [1.0, 0.0, 1.0]
data = [exp(M, p,  σ * rand(M; vector_at=p)) for i in 1:n];
f(M, p) = sum(1 / (2 * n) * distance.(Ref(M), Ref(p), data) .^ 2)
```

We can now generate the problem with its objective and the state

```{julia}
mp = DefaultManoptProblem(M, ManifoldCostObjective(f))
s = RandomWalkState(M; σ = 0.2)

solve!(mp, s)
get_solver_result(s)
```

The function `solve!` works also in place of `s`, but the last line illustrates how to access
the result in general; we could also just look at `s.p`, but the function `get_iterate` is
also used in several other places.

> We could for example easilz set up a second solver to work from a specified starting point

Reviewer: ```suggestion
We could for example easily set up a second solver to work from a specified starting point
```

Author: Oh, yes, I wrote some par t of that at home on my German keyboard, and the one thing to get consistent then is the switch of z and y on said keyboard ;)

given start point unchanged would just add a `copy(M, p)` upfront.
`Manopt.jl` provides both variants.

```{julia}
function random_walk_algorithm!(
    M::AbstractManifold,
    mgo::ManifoldCostObjective,
    p;
    σ = 0.1,
    retraction_method::AbstractRetractionMethod=default_retraction_method(M, typeof(p)),
    stopping_criterion::StoppingCriterion=StopAfterIteration(200),
    kwargs...,
)
    dmgo = decorate_objective!(M, mgo; kwargs...)
    dmp = DefaultManoptProblem(M, dmgo)
    s = RandomWalkState(M, [1.0, 0.0, 0.0];
        σ=0.1,
        retraction_method=retraction_method, stopping_criterion=stopping_criterion,
    )
    ds = decorate_state!(s; kwargs...)
    solve!(dmp, ds)
    return get_solver_return(get_objective(dmp), ds)
end
```

### The high level interface

Starting from the last section, the usual call a user would prefer is just
passing a manifold `M` the cost `f` and maybe a start point `p`.

```{julia}
function random_walk_algorithm!(M::AbstractManifold, f, p=rand(M); kwargs...)
    mgo = ManifoldCostObjective(f)
    return random_walk_algorithm!(M, mgo, p; kwargs...)
end
```

## Ease of Use II: The State Summary

> For the case that zou set `return_state=true` the soplver should return a summary of the run. This is realised providing a `show` method.

Reviewer: ```suggestion
For the case that you set `return_state=true` the solver should return a summary of the run. When a `show` method is provided, users can easily read such summary in a terminal.
```

> It should reflect its main parameters, if they are not too verbose an dprovide information

Reviewer: ```suggestion
It should reflect its main parameters, if they are not too verbose and provide information
```

> about the reason it stopped and whether this indicates convergence

Reviewer: ```suggestion
about the reason it stopped and whether this indicates convergence.
```
